### PR TITLE
adapter: remove session from more coord commands

### PR DESCRIPTION
**Pull request by mjibson**
Continue work to prepare for RPC-ability of the coordinator.

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

    * Reviewing each commit will be easiest.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Review by umanwizard — APPROVED** [review]

**Comment by umanwizard** [comment]
My approval is specifically for moving around the statement logging stuff. I haven't reviewed the rest of the code.

File: src/adapter/src/client.rs [hunk 1]
```diff
@@ -599,15 +601,26 @@ impl SessionClient {
...
+        // TODO: Remove this clone once we always have the session. It's currently needed because
+        // self.session returns a mut ref, so we can't call it twice.
+        let pcx = self.session().pcx().clone();
+        let conn_catalog = catalog.for_session(self.session());
+        let result: Result<_, AdapterError> =
+            mz_sql::plan::plan_copy_from(&pcx, &conn_catalog, id, columns, rows)
+                .err_into()
+                .and_then(|values| values.lower().err_into())
+                .and_then(|values| {
+                    Optimizer::logical_optimizer(&mz_transform::typecheck::empty_context())
```
Comment (andrewrodriguez-m) [inline src/adapter/src/client.rs:614]: Not super actionable so feel free to proceed regardless but it does seem strange for the "client" side to care about needing to optimize

**Review by andrewrodriguez-m — APPROVED** [review]

File: src/adapter/src/client.rs (hunk 1 above)
Comment (mjibson) [reply src/adapter/src/client.rs:614]: We either optimize in the coordinator or not in the coordinator, and we want to move as many things as possible out of the coordinator. https://github.com/MaterializeInc/materialize/pull/18994 was recently merged to move peek optimization out. This is another one of those.

**Comment by aalexandrov** [comment]
> [...] prepare for RPC-ability of the coordinator.

Is there a design doc for this? It would be great if I am aware and can stick to adapter conventions that are going to be introduced  imminently as I'm going over the `~optimize~` methods in #20652 and #18089.

**Comment by aalexandrov** [comment]
> [...] prepare for RPC-ability of the coordinator.

@mjibson Is there a design doc for this? It would be better if I am aware and can stick to / enforce adapter conventions that are going to be introduced imminently as I'm going over the `~optimize~` methods in #20652 and #18089.

**Comment by mjibson** [comment]
There's not a design document yet. It's a huge refactor and I believe any design doc would change many times during actual implementation. I may write one after the work is done so it's documented. Here's what you can do now:

1. If possible, only rely on a `ConnectionId` and/or a `Catalog`: these are both cheap data structures that we pass around to anything that needs them. Things that only need this can live fully in the adapter client.
2. Try to avoid passing a Session from the adapter client to the coordinator. It's fine if you do (I'll remove it eventually), but it'd be beneficial to not have even more of them. If you are adding a new Plan type or adding/changing a method that processes a plan (some sequence_x method called by sequence_plan), then don't worry about this at all and use the Session however you like. I'm going to change all of these and the work here won't affect me.

If you're doing optimization/explain stuff: it may be possible to do everything you need with a Catalog and a dataflow builder, both of which are RPC-safe. If you can stay away from using coordinator state or references then it'll be much easier to move anything here into the adapter client.

**Pull request merged**